src/main/scala/io/iohk/ethereum/network/p2p/FrameCodec.scala

```diff
@@ -0,0 +1,186 @@
+package io.iohk.ethereum.network.p2p
+
+import java.io.IOException
+
+import akka.util.ByteString
+import io.iohk.ethereum.network.Secrets
+import io.iohk.ethereum.utils.RLP
+import io.iohk.ethereum.utils.RLPImplicits._
+import org.spongycastle.crypto.StreamCipher
+import org.spongycastle.crypto.digests.KeccakDigest
+import org.spongycastle.crypto.engines.AESFastEngine
+import org.spongycastle.crypto.modes.SICBlockCipher
+import org.spongycastle.crypto.params.{KeyParameter, ParametersWithIV}
+
+case class Frame(header: Header, `type`: Int, payload: Array[Byte])
+
+case class Header(bodySize: Int, protocol: Int, contextId: Option[Int], totalFrameSize: Option[Int])
+
+class FrameCodec(private val secrets: Secrets) {
+
+  private val blockSize = secrets.aes.length * 8
+
+  private val enc: StreamCipher = {
+    val cipher = new SICBlockCipher(new AESFastEngine)
+    cipher.init(true, new ParametersWithIV(new KeyParameter(secrets.aes), new Array[Byte](blockSize / 8)))
+    cipher
+  }
+
+  private val dec: StreamCipher = {
+    val cipher = new SICBlockCipher(new AESFastEngine)
+    cipher.init(false, new ParametersWithIV(new KeyParameter(secrets.aes), new Array[Byte](blockSize / 8)))
+    cipher
+  }
+
+  private var unprocessedData: ByteString = ByteString.empty
+
+  private var headerOpt: Option[Header] = None
+
+  def readFrames(data: ByteString): Seq[Frame] = {
+    unprocessedData ++= data
+
+    def readRecursive(): Seq[Frame] = {
```

Comment:
@LukasGasior1 What about transforming into this function into tail recursive and using @tailrec annotation? It can be rapidly done using an accumulator as a parameter.